src/main/java/com/arpnetworking/metrics/mad/parsers/PrometheusToRecordParser.java

```diff
@@ -0,0 +1,245 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.metrics.mad.parsers;
+
+import com.arpnetworking.commons.builder.ThreadLocalBuilder;
+import com.arpnetworking.metrics.common.parsers.Parser;
+import com.arpnetworking.metrics.common.parsers.exceptions.ParsingException;
+import com.arpnetworking.metrics.mad.model.DefaultMetric;
+import com.arpnetworking.metrics.mad.model.DefaultRecord;
+import com.arpnetworking.metrics.mad.model.HttpRequest;
+import com.arpnetworking.metrics.mad.model.Metric;
+import com.arpnetworking.metrics.mad.model.Record;
+import com.arpnetworking.metrics.prometheus.Remote;
+import com.arpnetworking.metrics.prometheus.Types;
+import com.arpnetworking.metrics.prometheus.Types.TimeSeries;
+import com.arpnetworking.tsdcore.model.MetricType;
+import com.arpnetworking.tsdcore.model.Quantity;
+import com.arpnetworking.tsdcore.model.Unit;
+import com.google.common.base.MoreObjects;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.protobuf.InvalidProtocolBufferException;
+import net.sf.oval.exception.ConstraintsViolatedException;
+import org.xerial.snappy.Snappy;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.List;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+
+/**
+ * Parses the Prometheus protobuf binary protocol into records.
+ *
+ * @author Bruno Green (bruno dot green at gmail dot com)
+ */
+public class PrometheusToRecordParser implements Parser<List<Record>, HttpRequest> {
+    /**
+     * public constructor.
+     * @param interpretUnits specifies whether or not to interpret units.
+     */
+    public PrometheusToRecordParser(final boolean interpretUnits) {
+        _interpretUnits = interpretUnits;
+    }
+    /**
+     * Parses a unit from the name of a metric.
+     * Prometheus will, by default, add unit names to the end of a metric name.
+     * We want to parse that name and apply that unit to the metric.
+     * An unit suffix might be added to the name of the metric, we currently have a set of
+     * whitelisted suffixes that is most likely not exhaustive.
+     * For more information see: https://prometheus.io/docs/practices/naming/
+     * */
+    Optional<ParseResult> parseUnit(final String name) {
+        if (!_interpretUnits) {
+            return Optional.empty();
```

Comment:
Instead of returning `empty()` here or below where you do not find a new name, why not simplify the logic by always returning a `ParseResult` with the name to use (set to either the input name or an extract name) and an optional unit.  This would remove the `result.map` calls below as well as the `isPresent` logic on name. There would be fewer conditional branches and the code would be easier to follow in my opinion.